controller/tests/controller_server/csi_controller_server_test.py

```diff
@@ -689,14 +698,159 @@ def _test_create_volume_from_snapshot_error(self, storage_agent, array_type, arr
+        storage_agent.return_value = self.storage_agent
+        vol_id = "wwn1"
+        self.request.volume_content_source = self._get_source(vol_id, config.VOLUME_TYPE_NAME)
+        self.mediator.get_volume = Mock()
+        self.mediator.get_volume.return_value = utils.get_mock_mediator_response_volume(10, "vol", "wwn2", "a9k")
+        self.mediator.get_object_by_id = Mock()
+        self.mediator.get_object_by_id.return_value = utils.get_mock_mediator_response_volume(1000, vol_name,
+                                                                                              "wwn", "a9k")
+        self.mediator.copy_to_existing_volume_from_source = Mock()
+        array_type.return_value = "a9k"
+        self.servicer.CreateVolume(self.request, self.context)
+        self.assertEqual(self.context.code, grpc.StatusCode.OK)
+
+    @patch("controller.controller_server.csi_controller_server.detect_array_type")
+    @patch("controller.controller_server.csi_controller_server.get_agent")
+    def test_clone_volume_error_other_source(self, storage_agent, array_type):
+        storage_agent.return_value = self.storage_agent
+        vol_id = "wwn1"
+        vol_src_id = "wwn3"
+        self.request.volume_content_source = self._get_source(vol_id, config.VOLUME_TYPE_NAME)
+        self.mediator.get_volume = Mock()
+        self.mediator.get_volume.return_value = utils.get_mock_mediator_response_volume(10, "vol", "wwn2", "a9k",
+                                                                                        copy_source_id=vol_src_id)
+        array_type.return_value = "a9k"
+        self.servicer.CreateVolume(self.request, self.context)
+        self.assertEqual(self.context.code, grpc.StatusCode.INTERNAL)
+
+    @patch("controller.controller_server.csi_controller_server.detect_array_type")
+    @patch("controller.controller_server.csi_controller_server.get_agent")
+    def test_clone_volume_src_volume_not_found(self, storage_agent, array_type):
+        array_exception = array_errors.ObjectNotFoundError("")
+        self._test_clone_volume_error(storage_agent, array_type, array_exception,
+                                      grpc.StatusCode.NOT_FOUND)
+
+    def test_clone_volume_src_volume_invalid(self):
+        volume_content_source = self._get_source('vol_id', config.VOLUME_TYPE_NAME)
+        volume_content_source.volume.volume_id = 'invalid_volume_id'
+        self.request.volume_content_source = volume_content_source
+
+        self.servicer.CreateVolume(self.request, self.context)
+
+        self.assertEqual(self.context.code, grpc.StatusCode.NOT_FOUND)
+        self.assertTrue("invalid_volume_id" in self.context.details)
+
+    @patch("controller.controller_server.csi_controller_server.detect_array_type")
+    @patch("controller.controller_server.csi_controller_server.get_agent")
+    def test_clone_volume_target_volume_not_found(self, storage_agent, array_type):

```

Comment:
how is it different than `test_clone_volume_src_volume_not_found`? it looks like the same test minus an assert (so the tested code would act the same)

Reply:
you right, its redundant

Reply:
done

Reply:
what actually redundant is `rollback_called` (everywhere).
in order to void this from happening in the future, please remove `rollback_called` completely, and try to unify your `_test_clone_volume_error` with the existing `_test_create_volume_from_snapshot_error` (since it's mainly duplication).

Reply:
in addition, the code that is removed here tried to cover scenarios that are still not covered (exceptions from e.g. `get_volume`, `get_object_by_id`, etc...).
so please cover them

Reply:
is there a way I can see what exactly covered?

Reply:
`pytest --cov-report term-missing`. if you find a better way, please write it here

Reply:
done
add coverage==4.0.0 to requirments